connectors/sources/gitlab/datasource.py

```diff
@@ -0,0 +1,1040 @@
+        """Extract assignees data from work item widgets."""
+        for widget in work_item.widgets:
+            if isinstance(widget, WorkItemWidgetAssignees):
+                return widget.assignees
+        return PaginatedList[GitLabUser](nodes=[])
+
+    def _extract_widget_labels(
+        self, work_item: GitLabWorkItem
+    ) -> PaginatedList[GitLabLabel]:
+        """Extract labels data from work item widgets."""
+        for widget in work_item.widgets:
+            if isinstance(widget, WorkItemWidgetLabels):
+                return widget.labels
+        return PaginatedList[GitLabLabel](nodes=[])
+
+    def _extract_widget_discussions(
+        self, work_item: GitLabWorkItem
+    ) -> PaginatedList[GitLabDiscussion]:
+        """Extract discussions from work item widgets."""
+        for widget in work_item.widgets:
+            if isinstance(widget, WorkItemWidgetNotes):
+                return widget.discussions
+        # Return empty paginated list if not found
+        return PaginatedList[GitLabDiscussion](nodes=[])
+
+    def _extract_widget_hierarchy(
+        self, work_item: GitLabWorkItem
+    ) -> WorkItemWidgetHierarchy | None:
+        """Extract hierarchy info from work item widgets (for Epics)."""
+        for widget in work_item.widgets:
+            if isinstance(widget, WorkItemWidgetHierarchy):
+                return widget
+        return None
+
+    def _extract_widget_linked_items(
+        self, work_item: GitLabWorkItem
+    ) -> WorkItemWidgetLinkedItems | None:
+        """Extract linked items from work item widgets (for Epics/Issues)."""
+        for widget in work_item.widgets:
+            if isinstance(widget, WorkItemWidgetLinkedItems):
+                return widget
+        return None
+
+    T = TypeVar("T", bound=BaseModel)
+
+    async def _fetch_remaining_paginated_field(
+        self,
+        paginated_list: PaginatedList[T],
+        project_path: str,
+        iid: int,
+        field_name: str,
+        issuable_type: str,
+        model_class: Type[T],
+    ) -> None:
+        """Generic method to fetch remaining items for any paginated field.
+
+        Args:
+            paginated_list: The paginated list to append items to
+            project_path: Full path of the project
+            iid: Issue or MR internal ID
+            field_name: Name of field ('assignees', 'labels', 'discussions', 'reviewers', 'approvedBy')
+            issuable_type: Type of issuable ('issue' or 'mergeRequest')
+            model_class: Pydantic model class for validation
+        """
+        if paginated_list.page_info.has_next_page:
+            cursor = paginated_list.page_info.end_cursor
+            async for item in self.gitlab_client.fetch_remaining_field(
+                project_path, iid, field_name, issuable_type, cursor
+            ):
+                paginated_list.nodes.append(model_class.model_validate(item))
+
+    async def _fetch_remaining_assignees(
```

Comment:
Why do methods below mention `remaining`, while they do not seem to be "remaining`, but rather regular objects?

Reply:
The query initially fetches a limited number of sub-items for MRs (e.g., first 100 assignees). These sub-items contain pagination info - if pageInfo.hasNextPage is true, then all remaining pages are exhausted under these fetch_remaining methods. If there's only one page, the method does nothing.